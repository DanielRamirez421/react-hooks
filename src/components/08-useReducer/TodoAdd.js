import React from 'react';
import { useForm } from '../../hooks/useForm';
import PropTypes from 'prop-types';

export const TodoAdd = ({ handleAdd }) => {

  const [ { description }, handleInputChange, resetFormValues ] = useForm({ description: '' });

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim() <= 2) { return; };
    
    handleAdd({ 
      id: new Date().getTime(), 
      desc: description.trim(), 
      done: false  
    })
    resetFormValues();
  }

  return (
  <div className='col-5'>
    <h4>Add toDo</h4>
    <form 
      className='form-group text-right'
      onSubmit={ handleSubmit }>
      <div className='row'>
        <div className='col-8'>
          <input 
            className='form-control w-100'
            type='text'
            name='description'
            placeholder='Add Todo'
            autoComplete='off'
            value={ description }
            onChange={ handleInputChange }
          />
        </div>
        <div className='col-4'>
          <button 
            className='btn btn-outline-primary w-100'
            type='submit'
            >
              Add
            </button>
        </div>
      </div>
    </form>
  </div>
  );
};

TodoAdd.propTypes = {
  handleAdd: PropTypes.func.isRequired,
}