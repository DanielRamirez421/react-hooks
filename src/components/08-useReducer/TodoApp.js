import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hooks/useForm';
import { todoReducer } from './todoReducer';

 import './styles.css'

const init = () => {
  return JSON.parse(localStorage.getItem('toDos')) || [];
}


export const TodoApp = () => {

  const [ toDosState, dispatch ] = useReducer(todoReducer, [], init);
  const [ { description }, handleInputChange, resetValues ] = useForm({ description: '' });

  useEffect(() => localStorage.setItem('toDos', JSON.stringify( toDosState )), [ toDosState ]);
  
  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim() <= 2) { return; };
    
    const action = { 
      type: 'add', payload: {
        id: new Date().getTime(), 
        desc: description.trim(), 
        done: false 
      }
    };

    dispatch( action );
    resetValues();
  }

  const handleDelete = ( todoId ) => {
    const deleteAction = { type: 'delete', payload: todoId };
    dispatch( deleteAction );    
  }

  return (
  <div>
    <h3>TodoApp length: { toDosState.length }</h3>
    <hr />

    <ul className='list-group list-group-flush'>
      {
        toDosState.map((t, i) => 
          <div key={t.id} className='row'>
            <div className='col-10'>
              <li className='list-group-item text-center'>{ i+1 }. { t.desc }</li>
            </div>
            <div className='col-2'>
              <button 
                className='btn btn-danger w-100'
                onClick={ () => handleDelete( t.id ) }>Delete</button>  
            </div>
          </div>
        )
      }
    </ul>

    <hr />

    <h4>Add toDo</h4>
    <form 
      className='form-group text-right'
      onSubmit={ handleSubmit }>
      <div className='row justify-content-end'>
        <div className='col-5'>
          <input 
            className='form-control w-100'
            type='text'
            name='description'
            placeholder='Add Todo'
            autoComplete='off'
            value={ description }
            onChange={ handleInputChange }
          />
        </div>
        <div className='col-1'>
          <button 
            className='btn btn-outline-primary w-100'
            type='submit'>Add</button>
        </div>
      </div>
    </form>

  </div>
  );
};
