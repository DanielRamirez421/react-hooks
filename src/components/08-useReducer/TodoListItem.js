import React from 'react';

export const TodoListItem = ({ todo, handleToggle, handleDelete, index }) => {
  return (
    <div 
      className='row my-1'
      onClick={ () => handleToggle(todo.id) }
      role='button'>
      <div className='col-10'>
        <li className={`list-group-item text-center ${ todo.done && 'completed' }`}>{ index + 1 }. { todo.desc }</li>
      </div>
      <div className='col-2'>
        <button 
          className='btn btn-danger'
          onClick={ () => handleDelete( todo.id ) }>Delete</button>  
      </div>
    </div>
  );
};
