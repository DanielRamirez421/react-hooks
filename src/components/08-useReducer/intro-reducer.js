

const initialState = [{
  id: 1,
  todo: 'buy milk',
  done: false
}];

const newTodo = {
  id: 2,
  todo: 'learn react',
  done: false
}

const todoReducer = ( state = initialState, action ) => {

  if (action?.type === 'newTodo') {
    return [ ...state, action.payload ];
  }

  return state;
};

let toDo = todoReducer();

console.log('before add new todo', toDo);

const todoAction = {
  type: 'newTodo',
  payload: newTodo
}

toDo = todoReducer( toDo, todoAction );

console.log('after add new todo', toDo);
