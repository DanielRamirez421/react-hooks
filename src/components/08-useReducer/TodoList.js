import React from 'react';
import { TodoListItem } from './TodoListItem';
export const TodoList = ({ toDosState, handleToggle, handleDelete }) => {
  return (
  <div className='col-6'>
      <ul className='list-group list-group-flush'>
      { 
        toDosState.map((todo, i) => 
          <TodoListItem 
            todo={ todo } 
            handleToggle={ handleToggle } 
            handleDelete={ handleDelete } 
            index={ i }
            key={ todo.id } 
          />
        )
      }
      </ul>
  </div>
  );
};
