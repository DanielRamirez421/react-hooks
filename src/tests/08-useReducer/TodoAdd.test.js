import { shallow } from "enzyme";
import { TodoAdd } from "../../components/08-useReducer/TodoAdd";

describe('TodoAdd component tests', () => {

  const handleAdd = jest.fn();
  const wrapper = shallow(<TodoAdd handleAdd={ handleAdd }/>);

  test('should show TodoAddComponent', () => { 
    expect(wrapper).toMatchSnapshot();
  })

  test('should not call handleAddTodo', () => { 
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault(){} });
    expect(handleAdd).toHaveBeenCalledTimes(0);
  });

  test('should call handleAddTodo with an argument', () => { 
    const value = 'ExampleTodo';
    wrapper.find('input').simulate('change', { target: { value, name: 'description' } });

    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault(){} });

    expect( handleAdd ).toHaveBeenCalledTimes(1);
    expect( handleAdd ).toHaveBeenCalledWith( expect.any(Object) );
    expect( handleAdd ).toHaveBeenCalledWith( { id: expect.any(Number), desc: value, done: false } );
    expect( wrapper.find('input').prop('value') ).toBe('');
  });

});