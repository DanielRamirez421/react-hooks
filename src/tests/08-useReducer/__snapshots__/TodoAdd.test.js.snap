// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdd component tests should show TodoAddComponent 1`] = `
<div
  className="col-5"
>
  <h4>
    Add toDo
  </h4>
  <form
    className="form-group text-right"
    onSubmit={[Function]}
  >
    <div
      className="row"
    >
      <div
        className="col-8"
      >
        <input
          autoComplete="off"
          className="form-control w-100"
          name="description"
          onChange={[Function]}
          placeholder="Add Todo"
          type="text"
          value=""
        />
      </div>
      <div
        className="col-4"
      >
        <button
          className="btn btn-outline-primary w-100"
          type="submit"
        >
          Add
        </button>
      </div>
    </div>
  </form>
</div>
`;
